<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html>

<!-- ---------------------------------------------------------------------- -->
<head>
<meta http-equiv="content-type" content="text/html; charset=ISO-8859-1">
<title>
Cassis: Detection of genomic rearrangement breakpoints
</title>
</head>

<body>
<!-- ---------------------------------------------------------------------- -->
<h2><b>Cassis:</b> <i>Detection of genomic rearrangement breakpoints</i>
<br></h2>
<hr>
<h3>Using Cassis to identify breakpoints in the genomes of <i>Homo sapiens</i>
and <i>Mus musculus</i></h3>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The package <b>Cassis</b> was used for the
comparison of the genomes of the species <i>Homo sapiens</i> (reference
genome) and <i>Mus musculus</i>.</p>

<!-- ---------------------------------------------------------------------- -->
<h3>1 - Input data</h3>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;To execute this comparison, we obtained
all necessary information from the
<a href="http://www.ensembl.org/index.html">Ensembl Genome Brower</a> site
(<b>Build 57</b>).</p>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The <b>Build 57</b> was released in March
3rd, 2010 and contains information about the <i>Homo sapiens</i>
assembly <b>GRCh37</b> (released in February 2009) and about the <i>Mus
musculus</i> assembly <b>NCBI m37</b> (released in April 2007).
</p>

<ol>
  <li><b>FASTA sequences</b>
  <p>The FASTA sequences of the chromosomes 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
  12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22 and X of the <i>Homo sapiens</i>
  and of the chromosomes 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
  16, 17, 18, 19 and X of the <i>Mus musculus</i> were downloaded from the
  Ensembl FTP site:
  <ul>
    <li>
    <a href="ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna/">
    <i>Homo sapiens</i>
    </li>
    <li>
    <a href="ftp://ftp.ensembl.org/pub/current_fasta/mus_musculus/dna/">
    <i>Mus musculus</i></a>
    </li>
  </ul>
  <br>
  We downloaded the FASTA files which contains the chromosome sequences
  after being processed by RepeatMasker (dna_rm).<br>
  All files are renamed according to the pattern:
  <tt>[chromosome_name].fasta</tt>
  (Example: <tt>1.fasta</tt>, <tt>2.fasta</tt>, <tt>X.fasta</tt>, etc.)
  </p>


  <li><b>Orthologous genes</b>
  <p>The list of orthologous genes was produced with data obtained through the
  <a href="http://www.ensembl.org/biomart/index.html">Biomart</a> tool. We
  downloaded the list of all orthologous genes which can be found between
  <i>Homo sapiens</i> and <i>Mus musculus</i> and the list of all
  <i>Mus musculus</i> genes. We processed these two lists with the script
  <tt>util/createOne2OneTable.pl</tt> (available in the <b>Cassis</b> package)
  to produce a TSV (Tab Separated Values) file containing all one-to-one
  orthologous genes.<br>
  <a href="./HomoSapiensMusMusculus_genes.txt">Download the list of one-to-one
  orthologous genes</a>
  </p>
</ol>
</p>

<!-- ---------------------------------------------------------------------- -->
<br><h3>2 - Results</h3>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The list of <tt>15,047</tt> pairs of
one-to-one orthologous genes was processed by the package <b>Cassis</b>.</p>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;In the first step of the method, <b>Cassis</b>
identified a total of <tt>369</tt> breakpoints. In the second step, these
breakpoints were submitted to the segmentation process.</p>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Table 1 shows the minimum, maximum, mean and
median breakpoint length before and after the segmentation process. It
also shows the number of breakpoints which received status <tt>1</tt> or
<tt>0</tt> (respectively, passed or failed in the statistical test that
evaluates the significance of the segmentation).</p>

<center>
<b>Table 1: Minimum, maximum, mean and median breakpoint length before and
after the segmentation</b>
<table border="1" cellspacing="0" cellpadding="2">
<tr align="center">
  <td rowspan="2"><b>Description</b></td>
  <td rowspan="2"><b>N</b></td>
  <td colspan="4"><b>Before segmentation [length in bp]</b></td>
  <td colspan="4"><b>After segmentation  [length in bp]</b></td>
</tr>
<tr align="center">
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
</tr>
<tr align="center">
  <td>All breakpoints</td>
  <td><tt>369</tt></td>
  <td><tt>1</tt></td>
  <td><tt>32,752,838</tt></td>
  <td><tt>771,401</tt></td>
  <td><tt>244,568</tt></td>
  <td><tt>21</tt></td>
  <td><tt>5,133,352</tt></td>
  <td><tt>201,251</tt></td>
  <td><tt>52,986</tt></td>
</tr>
<tr align="center">
  <td>Status = 1 </td>
  <td><tt>365</tt></td>
  <td><tt>1</tt></td>
  <td><tt>32,752,838</tt></td>
  <td><tt>775,512</tt></td>
  <td><tt>244,568</tt></td>
  <td><tt>21</tt></td>
  <td><tt>5,133,352</tt></td>
  <td><tt>203,057</tt></td>
  <td><tt>53,262</tt></td>
</tr>
<tr align="center">
  <td>Status = 0</td>
  <td><tt>4</tt></td>
  <td><tt>79,048</tt></td>
  <td><tt>1,031,931</tt></td>
  <td><tt>396,243</tt></td>
  <td><tt>236,998</tt></td>
  <td><tt>2,069</tt></td>
  <td><tt>72,226</tt></td>
  <td><tt>36,459</tt></td>
  <td><tt>35.771</tt></td>
</tr>
</table>
</center>
<br>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Of the <tt>365</tt> breakpoints which were
classified with status <tt>1</tt>, a total of <tt>25</tt> breakpoints had
their length increased after the segmentation. Table 2 shows the minimum,
maximum, mean and median length of these breakpoints, before and after the
segmentation process.</p>

<center>
<b>Table 2: Minimum, maximum, mean and median length of the <tt>25</tt>
breakpoints which had their length increased after the segmentation</b>
<table border="1" cellspacing="0" cellpadding="2">
<tr align="center">
  <td rowspan="2"><b>Description</b></td>
  <td rowspan="2"><b>N</b></td>
  <td colspan="4"><b>Before segmentation [length in bp]</b></td>
  <td colspan="4"><b>After segmentation  [length in bp]</b></td>
</tr>
<tr align="center">
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
</tr>
<tr align="center">
  <td>Length (after) &ge; Length (before)</td>
  <td><tt>25</tt></td>
  <td><tt>1</tt></td>
  <td><tt>5,115,573</tt></td>
  <td><tt>282,827</tt></td>
  <td><tt>34,067</tt></td>
  <td><tt>1,539</tt></td>
  <td><tt>5,133,352</tt></td>
  <td><tt>311,471</tt></td>
  <td><tt>51,645</tt></td>
</tr>
</table>
</center>
<br>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A total of <tt>340</tt> breakpoints were
classified with status <tt>1</tt> and had their length decreased after the
segmentation process. Table 3 shows the minimum, maximum, mean and median
length of these breakpoints, before and after the segmentation process.</p>

<center>
<b>Table 3: Minimum, maximum, mean and median length of the <tt>340</tt>
breakpoints which had their length decreased after the segmentation</b>
<table border="1" cellspacing="0" cellpadding="2">
<tr align="center">
  <td rowspan="2"><b>Description</b></td>
  <td rowspan="2"><b>N</b></td>
  <td colspan="4"><b>Before segmentation [length in bp]</b></td>
  <td colspan="4"><b>After segmentation  [length in bp]</b></td>
  <td colspan="4"><b>Length reduction [%]</b></td>
</tr>
<tr align="center">
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
  <td><i>Min</i></td>
  <td><i>Max</i></td>
  <td><i>Mean</i></td>
  <td><i>Median</i></td>
</tr>
<tr align="center">
  <td>Length (after) < Length (before)</td>
  <td><tt>340</tt></td>
  <td><tt>303</tt></td>
  <td><tt>32,752,838</tt></td>
  <td><tt>811,739</tt></td>
  <td><tt>256,577</tt></td>
  <td><tt>21</tt></td>
  <td><tt>4,860,908</tt></td>
  <td><tt>195,085</tt></td>
  <td><tt>54,883</tt></td>
  <td><tt>0.31</tt></td>
  <td><tt>99.99</tt></td>
  <td><tt>62.26</tt></td>
  <td><tt>71.20</tt></td>
</tr>
</table>
</center>
<br>

<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Figure 1 shows the histogram of the
differences of the breakpoint lengths, before and after the segmentation step,
for the <tt>358</tt> breakpoints which were classified with status <tt>1</tt>.</p>

<center>
<b>Figure 1: Histogram of the breakpoint size differences between "before"
and "after" the refinement step of Cassis.<br>(7 values larger than 5 Mbp have
been removed for the purpose of readability)</b><br>
<img src="refinement-histogram-genes.png" border="1">
</center>

<!-- ---------------------------------------------------------------------- -->
<br><h3>3 - Download</h3>
<ul>
  <li><a href="NonRefinedBreakpoints_genes.txt">NonRefinedBreakpoints.txt</a>
  - List of breakpoints before the segmentation.
  <li><a href="segmentation_genes.txt">segmentation.txt</a>
  - List of breakpoints after the segmentation.
</ul>
</p>

</body>
</html>
